### ISSUE TITLE
[Question] Why only refork on Linux w/ PR_SET_CHILD_SUBREAPER?

### ISSUE THREAD

**Issue by jpcamara** [Shopify/pitchfork#17]
I'm curious why `PR_SET_CHILD_SUBREAPER` is a hard requirement for reforking in pitchfork.

From what I've read in the Puma implementation code and [documentation](https://github.com/puma/puma/blob/master/docs/fork_worker.md), reforking there can happen on any unix-like system. It doesn't seem to require or utilize subreaping.

In pitchfork, if `REFORKING_AVAILABLE` is `false` no reforking happens at all and it's just a multiprocess preloading server. I can guess at some benefits of this approach:

- Some kind of issues with nesting? 
- Ability to create molds of future generations that maybe isn't possible in the non-subreaping approach?

But the documentation and code mostly tells me what the requirements are, but not why they are a necessity. I'm looking into writing a generalized reforking worker pool, so i'm curious what research led to a subreaper focused implementation.

Thanks!

**Comment by casperisfine** [comment]
Yes, `PR_SET_CHILD_SUBREAPER` is a requirement because otherwise the reforked workers are grand-children of the master process:

```
- master
  \ - worker-0
     \ - worker-1
     | - worker-2
     | - worker-3
```

The problems with this are:

  - If `worker-0` die, its children are reparented to the init process (pid 1) and the master may no longer be allowed to signal them.
  - This architecture either limit you to a single refork event, or force you to always refork from worker 0, meaning you need to let worker 0 continue to process user requests. If worker 0 acts both as a "mold" and a "worker" serving traffic, then:
    - you can no longer safely kill it on timeout
    - if you need to respawn another worker, you have to wait for worker-0 to finish its current request.

> I'm looking into writing a generalized reforking worker poo

Yeah that would be cool. I toyed with this idea (e.g. having some kind of `pitchfork-core` gem) during the implementation. We may get there eventually as we'll likely want to make our job processor use the same technique. But short term I was more confident validating the idea using a battle tested base like Unicorn.

IMO the key properties for reforking are:

  - The mold need to be inactive. It should wait for the master commands at all time and never process units of work (request or job).
  - This means you must be able to ask a worker to stop processing work in a timely manner (e.g. request or job timeout). Otherwise you could end up waiting an eternity to promote a worker and it would get messy.

So if you were to implement a generic reforking process pool library, the applicative layer using it (e.g. Sidekiq) would need to respect these principles.

Closing the issue, but if some things are unclear I can elaborate more.
